src/protocols/der/decode.c

```diff
@@ -0,0 +1,2852 @@
+		['5'] = true, ['6'] = true,  ['7'] = true, ['8'] = true, ['9'] = true,	     [':'] = true, ['='] = true,
+		['?'] = true, ['A'] = true,  ['B'] = true, ['C'] = true, ['D'] = true,	     ['E'] = true, ['F'] = true,
+		['G'] = true, ['H'] = true,  ['I'] = true, ['J'] = true, ['K'] = true,	     ['L'] = true, ['M'] = true,
+		['N'] = true, ['O'] = true,  ['P'] = true, ['Q'] = true, ['R'] = true,	     ['S'] = true, ['T'] = true,
+		['U'] = true, ['V'] = true,  ['W'] = true, ['X'] = true, ['Y'] = true,	     ['Z'] = true, ['a'] = true,
+		['b'] = true, ['c'] = true,  ['d'] = true, ['e'] = true, ['f'] = true,	     ['g'] = true, ['h'] = true,
+		['i'] = true, ['j'] = true,  ['k'] = true, ['l'] = true, ['m'] = true,	     ['n'] = true, ['o'] = true,
+		['p'] = true, ['q'] = true,  ['r'] = true, ['s'] = true, ['t'] = true,	     ['u'] = true, ['v'] = true,
+		['w'] = true, ['x'] = true,  ['y'] = true, ['z'] = true, [UINT8_MAX] = false
+	};
+
+	size_t len = fr_dbuff_remaining(&our_in);
+
+	if (!fr_type_is_string(parent->type)) {
+		fr_strerror_printf("Printable string found in non-string attribute %s of type %s", parent->name,
+				   fr_type_to_str(parent->type));
+		return -1;
+	}
+
+	/*
+	 *	ISO/IEC 8825-1:2021
+	 * 	8.23 Encoding for values of the restricted character string types
+	 *	8.23.1 The data value consists of a string of characters from the character set specified in the ASN.1
+	 *		type definition. 8.23.2 Each data value shall be encoded independently of other data values of
+	 *		the same type.
+	 *	8.23.3 Each character string type shall be encoded as if it had been declared:
+	 *			[UNIVERSAL x] IMPLICIT OCTET STRING
+	 *		where x is the number of the universal class tag assigned to the character string type in
+	 *		Rec. ITU-T X.680 | ISO/IEC 8824-1. The value of the octet string is specified in 8.23.4 and
+	 *		8.23.5.
+	 */
+
+	vp = fr_pair_afrom_da(ctx, parent);
+	if (unlikely(vp == NULL)) {
+		fr_strerror_const("Out of memory for printable string pair");
+		return -1;
+	}
+
+	if (unlikely(fr_pair_value_bstr_alloc(vp, &str, len, false) < 0)) {
+		fr_strerror_const("Out of memory for printable string");
+		return -1;
+	}
+
+	fr_dbuff_out_memcpy((uint8_t *)str, &our_in, len);
+
+	for (size_t i = 0; i < len; i++) {
+		/*
+		 *	Check that the byte is a printable ASCII character allowed in a printable string
+		 */
+		if (allowed_chars[(uint8_t)str[i]] == false) {
+			fr_strerror_printf("Invalid character in printable string (%d)", str[i]);
+			return -1;
+		}
+	}
+
+	str[len] = '\0';
+
+	fr_pair_append(out, vp);
+
+	return fr_dbuff_set(in, &our_in);
+}
+
+static ssize_t fr_der_decode_t61_string(TALLOC_CTX *ctx, fr_pair_list_t *out, fr_dict_attr_t const *parent,
+					fr_dbuff_t *in, UNUSED fr_der_decode_ctx_t *decode_ctx)
+{
+	fr_pair_t *vp;
+	fr_dbuff_t our_in = FR_DBUFF(in);
+	char	  *str	  = NULL;
+
+	static bool const allowed_chars[] = {
+		[0x08] = true, [0x0A] = true, [0x0C] = true,	  [0x0D] = true, [0x0E] = true, [0x0F] = true,
```

Comment:
see `src/lib/util/token.c`.  GCC and clang support ranges, which makes these tables a _lot_ easier

```
/*
 *	This is fine.  Don't complain.
 */
#ifdef __clang__
#pragma clang diagnostic ignored "-Wgnu-designator"
#endif
...
	[ 0 ... T_HASH ] = '?',	/* GCC extension for range initialization, also allowed by clang */
```